Document Uncommitted GC (#4670)
treeverse/lakeFS: docs/reference/garbage-collection.md

1. Objects are deleted or overwritten.
2. Changes are reverted from a branch before being committed.
3. A branch with uncommitted changes is deleted.

The uncommitted GC will not clean:
1. Objects that are included in any commit. For committed objects cleanup see [above](#what-gets-collected)
2. Everything mentioned in [what does not get collected](#what-does-_not_-get-collected)

To run the uncommitted GC job run:
  ```bash
spark-submit \
    --conf spark.hadoop.lakefs.api.url<LAKEFS_ENDPOINT> \
    --conf spark.hadoop.fs.s3a.access.key=<AWS_ACCESS_KEY_ID> \
    --conf spark.hadoop.fs.s3a.secret.key=<AWS_SECRET_ACCESS_KEY> \
    --conf spark.hadoop.lakefs.api.access_key=<LAKEFS_ACCESS_KEY_ID> \
    --conf spark.hadoop.lakefs.api.secret_key=<LAKEFS__SECRET_ACCESS_KEY> \
    --class io.treeverse.gc.UncommittedGarbageCollector \
    --packages org.apache.hadoop:hadoop-aws:2.7.7,\
          lakefs-spark-client-<VERSION>-assembly-<TODO>.jar \
    <REPO_NAME> <TODO_OUTPUT_LOCATION>
```

#### Uncommitted GC job options

Similar to the described [above](#gc-job-options).

#### Backup and restore 

> <TODO> - will it work?

Reviewer: It will work, but lets put a pin on that for the first milestone

Author: Just like mark&sweep - it's a safety mechanisms that users can rely on for safer delete. It can wait though.

Reviewer: So let's change the comment to 'be supported in the future'

#### Limitations
{: .no_toc }

The uncommitted GC job has several limitations in its Beta version: 
1. No writes to lakeFS during the execution of the job. Objects written to lakeFS

> during the job run may or may not be detected by the job. It's advised to avoid any

Reviewer: I think we should be more strict than "advise". More on the lines of any non-read operation during execution of the beta GC process might result in unexpected behavior

Author: Done.

write operation while the job is running, like `UploadObject`, `CopyObject`, `StageObject`, 
`LinkPhysicalAddress` or any other non-read operation.
2. Only available for S3 repositories at the moment.
3. Scale may be limited, see performance results below.

#### Next steps

The uncommitted GC is under development, next releases will include:

1. Incorporation of committed & uncommitted GC into a single job. We understand the friction
of having 2 garbage collection jobs for a lakeFS installation and working to creating a
single job for it.

> 2. Support for non-S3 repositories.

Reviewer: In my opinion this should be at the bottom of the list

Author: You got it.

  objects-to-restore-list-location backup-external-location storage-namespace azure
  ```
</div>

## Beta: Deleting uncommitted objects

> Deletion of objects that were never committed was always a difficulty for lakeFS, see
> [#1933](https://github.com/treeverse/lakeFS/issues/1933) for more details. Examples for 
> objects that will be collected as part of the uncommitted GC job:

Reviewer 2: ```suggestion
Run this job to [delete objects that were never committed and have no valid pointers from lakeFS](https://github.com/treeverse/lakeFS/issues/1933). This may happen when:
```

> 1. Objects that were uploaded to lakeFS and deleted.
> 2. Objects that were uploaded to lakeFS and were overridden.

Reviewer 2: ```suggestion
1. Objects are deleted or overwritten.
2. Changes are reverted from a branch before being committed.
3. A branch with uncommitted changes is deleted.
```

The uncommitted GC will not clean:

> 1. Committed objects. For committed objects cleanup see [above](#what-gets-collected)

Reviewer 2: ```suggestion
1. Objects that are included in any commit. For committed objects cleanup see [above](#what-gets-collected)
```
